File: packages/ce-core/src/cost.ts
/**
 * Cost Estimation with Cache Savings
 *
 * Estimates actual API costs for context packs, with special support
 * for prefix caching savings. Given a CacheAwarePack, shows concrete
 * dollar amounts saved by cache-topology-aware packing.
 *
 * Pricing as of 2025-06 (per million tokens):
 * - Claude Opus 4.6:   $15 input, $1.50 cached input, $75 output
 * - Claude Sonnet 4.6: $3 input, $0.30 cached input, $15 output
 * - GPT-4.1:          $2 input, $0.50 cached input, $8 output
 * - GPT-4o:           $2.50 input, $1.25 cached input, $10 output
 */

import type { CacheAwarePack } from "./cache-topology.js";
import { ValidationError } from "./errors.js";

/**
 * Pricing per million tokens for a model.
 */
export interface ModelPricing {
  /** Cost per million input tokens */
  inputPerMillion: number;
  /** Cost per million cached input tokens (prefix cache hit) */
  cachedInputPerMillion: number;
  /** Cost per million output tokens */
  outputPerMillion: number;
}

/**
 * Cost estimate for a single request.
 */
export interface CostEstimate {
  /** Model used for pricing */
  model: string;
  /** Total input tokens */
  inputTokens: number;
  /** Tokens that hit the prefix cache */
  cachedTokens: number;
  /** Tokens that miss the cache (billed at full rate) */
  uncachedTokens: number;
  /** Estimated output tokens (for total cost) */
  outputTokens: number;
  /** Cost without caching (all tokens at full input price) */
  costWithoutCache: number;
  /** Cost with caching (cached tokens at reduced rate) */
  costWithCache: number;
  /** Dollar savings from caching */
  savings: number;
  /** Savings as percentage */
  savingsPercent: number;
  /** Cache efficiency (0-1) */
  cacheEfficiency: number;
}

/**
 * Cost projection over multiple requests.
 */
export interface CostProjection {
  /** Per-request estimates */
  perRequest: CostEstimate;
  /** Number of requests projected */
  requestCount: number;
  /** Total cost without caching */
  totalWithoutCache: number;
  /** Total cost with caching */
  totalWithCache: number;
  /** Total savings */
  totalSavings: number;
  /** Monthly projection at this request rate */
  monthlyEstimate?: {
    requestsPerDay: number;
    monthlyCostWithoutCache: number;
    monthlyCostWithCache: number;
    monthlySavings: number;
  };
}

/**
 * Known model pricing (approximate, check provider docs for current rates).
 */
export const MODEL_PRICING: Record<string, ModelPricing> = {
  // Anthropic
  "claude-opus-4-6": {
    inputPerMillion: 15,
    cachedInputPerMillion: 1.5,
    outputPerMillion: 75,
  },
  "claude-sonnet-4-6": {
    inputPerMillion: 3,
    cachedInputPerMillion: 0.3,
    outputPerMillion: 15,
  },
  "claude-haiku-4-5": {
    inputPerMillion: 0.8,
    cachedInputPerMillion: 0.08,
    outputPerMillion: 4,
  },

  // OpenAI
  "gpt-4.1": {
    inputPerMillion: 2,
    cachedInputPerMillion: 0.5,
    outputPerMillion: 8,
  },
  "gpt-4.1-mini": {
    inputPerMillion: 0.4,
    cachedInputPerMillion: 0.1,
    outputPerMillion: 1.6,
  },
  "gpt-4o": {
    inputPerMillion: 2.5,
    cachedInputPerMillion: 1.25,
    outputPerMillion: 10,
  },
  "o3": {
    inputPerMillion: 2,
    cachedInputPerMillion: 0.5,
    outputPerMillion: 8,
  },
  "o4-mini": {
    inputPerMillion: 1.1,
    cachedInputPerMillion: 0.275,
    outputPerMillion: 4.4,
  },
};

/**
 * Estimate the cost of a single API request using a cache-aware pack.
 *
 * @param pack - Cache-aware pack result from packWithCacheTopology
 * @param model - Model name for pricing lookup
 * @param outputTokens - Estimated output tokens (default: 500)
 * @param pricing - Custom pricing (overrides MODEL_PRICING lookup)
 *
 * @example
 * ```ts
 * const pack = packWithCacheTopology(items, budget, {}, { provider: "anthropic" });
 * const cost = estimateCost(pack, "claude-sonnet-4-6");
 * console.log(`Saving $${cost.savings.toFixed(4)} per request (${cost.savingsPercent.toFixed(1)}%)`);
 * ```
 */
export function estimateCost(
  pack: CacheAwarePack,
  model: string,
  outputTokens: number = 500,
  pricing?: ModelPricing,
): CostEstimate {
  const price = pricing ?? MODEL_PRICING[model];
  if (!price) {
    const known = Object.keys(MODEL_PRICING).join(", ");
    throw new ValidationError(
      `Unknown model "${model}". Known models: ${known}. Pass custom pricing for unlisted models.`,
      [{ path: "model", message: `Expected one of: ${known}` }],
    );
  }

  const cachedTokens = pack.cacheableTokens;
  const uncachedTokens = pack.volatileTokens;
  const totalInput = pack.totalTokens;

  const costWithoutCache =
    (totalInput / 1_000_000) * price.inputPerMillion +
    (outputTokens / 1_000_000) * price.outputPerMillion;

  const costWithCache =
    (cachedTokens / 1_000_000) * price.cachedInputPerMillion +
    (uncachedTokens / 1_000_000) * price.inputPerMillion +
    (outputTokens / 1_000_000) * price.outputPerMillion;

  const savings = costWithoutCache - costWithCache;

  return {
    model,
    inputTokens: totalInput,
    cachedTokens,
    uncachedTokens,
    outputTokens,
    costWithoutCache: Math.round(costWithoutCache * 1_000_000) / 1_000_000,
    costWithCache: Math.round(costWithCache * 1_000_000) / 1_000_000,
    savings: Math.round(savings * 1_000_000) / 1_000_000,
    savingsPercent: costWithoutCache > 0
      ? Math.round((savings / costWithoutCache) * 1000) / 10
      : 0,
    cacheEfficiency: pack.cacheEfficiency,
  };
}

/**
 * Project costs over multiple requests.
 *
 * @param pack - Cache-aware pack result
 * @param model - Model name
 * @param requestCount - Number of requests to project
 * @param options - Additional options (outputTokens, requestsPerDay for monthly)
 *
 * @example
 * ```ts
 * const projection = projectCosts(pack, "claude-sonnet-4-6", 1000, {
 *   requestsPerDay: 500,
 * });
 * console.log(`Monthly savings: $${projection.monthlyEstimate?.monthlySavings.toFixed(2)}`);
 * ```
 */
export function projectCosts(
  pack: CacheAwarePack,
  model: string,
  requestCount: number,
  options?: {
    outputTokens?: number;
    pricing?: ModelPricing;
    requestsPerDay?: number;
  },
): CostProjection {
  const perRequest = estimateCost(
    pack,
    model,
    options?.outputTokens ?? 500,
    options?.pricing,
  );

  const totalWithoutCache = perRequest.costWithoutCache * requestCount;
  const totalWithCache = perRequest.costWithCache * requestCount;

  const result: CostProjection = {
    perRequest,
    requestCount,
    totalWithoutCache: Math.round(totalWithoutCache * 100) / 100,
    totalWithCache: Math.round(totalWithCache * 100) / 100,
    totalSavings: Math.round((totalWithoutCache - totalWithCache) * 100) / 100,
  };

  if (options?.requestsPerDay) {
    const monthlyRequests = options.requestsPerDay * 30;
    result.monthlyEstimate = {
      requestsPerDay: options.requestsPerDay,
      monthlyCostWithoutCache: Math.round(perRequest.costWithoutCache * monthlyRequests * 100) / 100,
      monthlyCostWithCache: Math.round(perRequest.costWithCache * monthlyRequests * 100) / 100,
      monthlySavings: Math.round(perRequest.savings * monthlyRequests * 100) / 100,
    };
  }

  return result;
}

File: python/context_engineering/cost.py
"""
Cost Estimation with Cache Savings

Estimates actual API costs for context packs, with special support
for prefix caching savings. Given a CacheAwarePack, shows concrete
dollar amounts saved by cache-topology-aware packing.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, Optional

from .cache_topology import CacheAwarePack


@dataclass
class ModelPricing:
    """Pricing per million tokens for a model."""
    input_per_million: float
    cached_input_per_million: float
    output_per_million: float


@dataclass
class CostEstimate:
    """Cost estimate for a single request."""
    model: str
    input_tokens: int
    cached_tokens: int
    uncached_tokens: int
    output_tokens: int
    cost_without_cache: float
    cost_with_cache: float
    savings: float
    savings_percent: float
    cache_efficiency: float


@dataclass
class MonthlyEstimate:
    requests_per_day: int
    monthly_cost_without_cache: float
    monthly_cost_with_cache: float
    monthly_savings: float


@dataclass
class CostProjection:
    """Cost projection over multiple requests."""
    per_request: CostEstimate
    request_count: int
    total_without_cache: float
    total_with_cache: float
    total_savings: float
    monthly_estimate: Optional[MonthlyEstimate] = None


MODEL_PRICING: Dict[str, ModelPricing] = {
    # Anthropic
    "claude-opus-4-6": ModelPricing(15, 1.5, 75),
    "claude-sonnet-4-6": ModelPricing(3, 0.3, 15),
    "claude-haiku-4-5": ModelPricing(0.8, 0.08, 4),
    # OpenAI
    "gpt-4.1": ModelPricing(2, 0.5, 8),
    "gpt-4.1-mini": ModelPricing(0.4, 0.1, 1.6),
    "gpt-4o": ModelPricing(2.5, 1.25, 10),
    "o3": ModelPricing(2, 0.5, 8),
    "o4-mini": ModelPricing(1.1, 0.275, 4.4),
}


def estimate_cost(
    pack: CacheAwarePack,
    model: str,
    output_tokens: int = 500,
    pricing: Optional[ModelPricing] = None,
) -> CostEstimate:
    """Estimate the cost of a single API request using a cache-aware pack.

    Args:
        pack: Cache-aware pack result from pack_with_cache_topology
        model: Model name for pricing lookup
        output_tokens: Estimated output tokens (default: 500)
        pricing: Custom pricing (overrides MODEL_PRICING lookup)

    Example:
        pack = pack_with_cache_topology(items, budget)
        cost = estimate_cost(pack, "claude-sonnet-4-6")
        print(f"Saving ${cost.savings:.4f} per request ({cost.savings_percent:.1f}%)")
    """
    price = pricing or MODEL_PRICING.get(model)
    if not price:
        models = ", ".join(MODEL_PRICING.keys())
        raise ValueError(f'Unknown model "{model}". Pass custom pricing or use one of: {models}')

    cached = pack.cacheable_tokens
    uncached = pack.volatile_tokens

    cost_without = (
        (pack.total_tokens / 1_000_000) * price.input_per_million
        + (output_tokens / 1_000_000) * price.output_per_million
    )
    cost_with = (
        (cached / 1_000_000) * price.cached_input_per_million
        + (uncached / 1_000_000) * price.input_per_million
        + (output_tokens / 1_000_000) * price.output_per_million
    )

    savings = cost_without - cost_with

    return CostEstimate(
        model=model,
        input_tokens=pack.total_tokens,
        cached_tokens=cached,
        uncached_tokens=uncached,
        output_tokens=output_tokens,
        cost_without_cache=round(cost_without, 6),
        cost_with_cache=round(cost_with, 6),
        savings=round(savings, 6),
        savings_percent=round((savings / cost_without) * 100, 1) if cost_without > 0 else 0,
        cache_efficiency=pack.cache_efficiency,
    )


def project_costs(
    pack: CacheAwarePack,
    model: str,
    request_count: int,
    output_tokens: int = 500,
    requests_per_day: Optional[int] = None,
    pricing: Optional[ModelPricing] = None,
) -> CostProjection:
    """Project costs over multiple requests.

    Args:
        pack: Cache-aware pack result
        model: Model name
        request_count: Number of requests to project
        output_tokens: Estimated output tokens per request
        requests_per_day: If set, includes monthly projection
        pricing: Custom pricing

    Example:
        projection = project_costs(pack, "claude-sonnet-4-6", 1000, requests_per_day=500)
        print(f"Monthly savings: ${projection.monthly_estimate.monthly_savings:.2f}")
    """
    per_request = estimate_cost(pack, model, output_tokens, pricing)

    total_without = per_request.cost_without_cache * request_count
    total_with = per_request.cost_with_cache * request_count

    monthly = None
    if requests_per_day is not None:
        monthly_requests = requests_per_day * 30
        monthly = MonthlyEstimate(
            requests_per_day=requests_per_day,
            monthly_cost_without_cache=round(per_request.cost_without_cache * monthly_requests, 2),
            monthly_cost_with_cache=round(per_request.cost_with_cache * monthly_requests, 2),
            monthly_savings=round(per_request.savings * monthly_requests, 2),
        )

    return CostProjection(
        per_request=per_request,
        request_count=request_count,
        total_without_cache=round(total_without, 2),
        total_with_cache=round(total_with, 2),
        total_savings=round(total_without - total_with, 2),
        monthly_estimate=monthly,
    )
